File: Backend/src/main/java/com/example/model/Video.java
package com.example.model;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;

@Entity
@Table(name = "videos")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Video {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, unique = true)
    private String youtubeId; // ID unique de YouTube (ex: "dQw4w9WgXcQ")

    @Column(nullable = false)
    private String title;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(nullable = false)
    private String thumbnailUrl; // URL de la miniature

    @Column(nullable = false)
    private String channelTitle; // Nom de la chaîne YouTube

    @Column(nullable = false)
    private Integer duration; // Durée en secondes

    @Column(nullable = false)
    private String category; // Mathématiques, Physique, etc.

    @Column(nullable = false)
    private String difficulty; // Facile, Moyen, Difficile

    @Column(nullable = false)
    private Integer viewCount = 0;

    @Column(nullable = false)
    private Integer favoriteCount = 0;

    @Column
    private String tags; // Tags séparés par des virgules

    @Column(nullable = false)
    @Builder.Default
    private Boolean isActive = true;

    @Column(nullable = false)
    @Builder.Default
    private Boolean isFeatured = false; // Vidéo mise en avant

    @CreationTimestamp
    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(nullable = false)
    private LocalDateTime updatedAt;

    // Méthode utilitaire pour formater la durée
    public String getFormattedDuration() {
        int hours = duration / 3600;
        int minutes = (duration % 3600) / 60;
        int seconds = duration % 60;

        if (hours > 0) {
            return String.format("%d:%02d:%02d", hours, minutes, seconds);
        } else {
            return String.format("%d:%02d", minutes, seconds);
        }
    }

    // Convertir les tags en liste
    public String[] getTagsList() {
        if (tags == null || tags.isEmpty()) {
            return new String[0];
        }
        return tags.split(",");
    }
}

File: Frontend/lib/models/video.dart
class Video {
  final int id;
  final String youtubeId;
  final String title;
  final String description;
  final String thumbnailUrl;
  final String channelTitle;
  final int duration;
  final String formattedDuration;
  final String category;
  final String difficulty;
  final int viewCount;
  final int favoriteCount;
  final List<String> tags;
  final bool isFavorite;
  final bool isWatched;
  final double progressPercentage;
  final int lastTimestamp;
  final DateTime createdAt;

  Video({
    required this.id,
    required this.youtubeId,
    required this.title,
    required this.description,
    required this.thumbnailUrl,
    required this.channelTitle,
    required this.duration,
    required this.formattedDuration,
    required this.category,
    required this.difficulty,
    required this.viewCount,
    required this.favoriteCount,
    required this.tags,
    required this.isFavorite,
    required this.isWatched,
    required this.progressPercentage,
    required this.lastTimestamp,
    required this.createdAt,
  });

  factory Video.fromJson(Map<String, dynamic> json) {
    return Video(
      id: json['id'],
      youtubeId: json['youtubeId'],
      title: json['title'],
      description: json['description'] ?? '',
      thumbnailUrl: json['thumbnailUrl'],
      channelTitle: json['channelTitle'],
      duration: json['duration'],
      formattedDuration: json['formattedDuration'],
      category: json['category'],
      difficulty: json['difficulty'],
      viewCount: json['viewCount'],
      favoriteCount: json['favoriteCount'],
      tags: List<String>.from(json['tags'] ?? []),
      isFavorite: json['isFavorite'],
      isWatched: json['isWatched'],
      progressPercentage: (json['progressPercentage'] ?? 0).toDouble(),
      lastTimestamp: json['lastTimestamp'] ?? 0,
      createdAt: DateTime.parse(json['createdAt']),
    );
  }

  // Add this copyWith method
  Video copyWith({
    int? id,
    String? youtubeId,
    String? title,
    String? description,
    String? thumbnailUrl,
    String? channelTitle,
    int? duration,
    String? formattedDuration,
    String? category,
    String? difficulty,
    int? viewCount,
    int? favoriteCount,
    List<String>? tags,
    bool? isFavorite,
    bool? isWatched,
    double? progressPercentage,
    int? lastTimestamp,
    DateTime? createdAt,
  }) {
    return Video(
      id: id ?? this.id,
      youtubeId: youtubeId ?? this.youtubeId,
      title: title ?? this.title,
      description: description ?? this.description,
      thumbnailUrl: thumbnailUrl ?? this.thumbnailUrl,
      channelTitle: channelTitle ?? this.channelTitle,
      duration: duration ?? this.duration,
      formattedDuration: formattedDuration ?? this.formattedDuration,
      category: category ?? this.category,
      difficulty: difficulty ?? this.difficulty,
      viewCount: viewCount ?? this.viewCount,
      favoriteCount: favoriteCount ?? this.favoriteCount,
      tags: tags ?? this.tags,
      isFavorite: isFavorite ?? this.isFavorite,
      isWatched: isWatched ?? this.isWatched,
      progressPercentage: progressPercentage ?? this.progressPercentage,
      lastTimestamp: lastTimestamp ?? this.lastTimestamp,
      createdAt: createdAt ?? this.createdAt,
    );
  }
}